internal/controller/servermaintenance_controller.go

```diff
@@ -348,22 +422,295 @@ func (r *ServerMaintenanceReconciler) cleanup(ctx context.Context, log logr.Logg
+			// we remove only the invalid boot options and keep the rest in same order
+			// because the settings will take care of new added boot device and apend it to end
+			currentBootDeviceMap := make(map[string]struct{})
+			for _, bootDevice := range bootOrder {
+				currentBootDeviceMap[bootDevice] = struct{}{}
+			}
+			var sanitizedBootOrder []string
+			for _, bootDevice := range serverMaintenance.Status.BootOrderStatus.DefaultBootOrder {
+				if _, ok := currentBootDeviceMap[bootDevice]; !ok {
+					log.V(1).Info("Removing invalid boot option from default boot order",
+						"Server", server.Name,
+						"BootDevice", bootDevice)
+					continue
+				}
+				sanitizedBootOrder = append(sanitizedBootOrder, bootDevice)
+			}
+
+			err = bmcClient.SetBootOrder(ctx, server.Spec.SystemURI, redfish.Boot{
+				BootOrder: sanitizedBootOrder,
+			})
+			if err != nil {
+				return false, fmt.Errorf("failed to revert boot order: %w", err)
+			}
+			// save the current boot order to revert back after maintenance
+			status := &metalv1alpha1.ServerMaintenanceBootOrder{
+				DefaultBootOrder: serverMaintenance.Status.BootOrderStatus.DefaultBootOrder,
+				State:            metalv1alpha1.BootOrderConfigSuccessRevertInProgress,
+			}
+			if err := r.patchBootOrderStatus(ctx, log, serverMaintenance, status); err != nil {
+				return false, fmt.Errorf("failed to patch maintenance default boot order: %w", err)
+			}
+			log.V(1).Info("Patched revert to default boot order", "Server", server.Name, "BootOrder", nil)
+			// requeue to give time to the server to actually do the work and verify it
+			return true, nil
+		}
+
+		if serverMaintenance.Status.BootOrderStatus.State == metalv1alpha1.BootOrderConfigSuccessRevertInProgress {
+			if server.Spec.Power != metalv1alpha1.PowerOn || server.Status.PowerState != metalv1alpha1.ServerOnPowerState {
+				if err := r.setAndPatchServerPowerState(ctx, log, server, metalv1alpha1.PowerOn); err != nil {
+					return false, err
+				}
+				// requeue to give time to the server to power on
+				return true, nil
+			}
+
+			revertCompleted := false
+			if reflect.DeepEqual(bootOrder, serverMaintenance.Status.BootOrderStatus.DefaultBootOrder) {
+				revertCompleted = true
+			} else {
+				// sometimes, changing biosSettings leads to change in number of boot options
+				// hence, we need check the default boot order and ignore extra from current boot order
+				for idx, bootDevice := range serverMaintenance.Status.BootOrderStatus.DefaultBootOrder {
+					// cases where boot option is disabled in bios settings
+					if idx >= len(bootOrder) {
+						break
+					}
+					if bootDevice != bootOrder[idx] {
+						log.V(1).Info("boot order for server is not yet reverted to default. Waiting...",
+							"Server", server.Name,
+							"CurrentBootOrder", bootOrder,
+							"DefaultBootOrder", serverMaintenance.Status.BootOrderStatus.DefaultBootOrder)
+						return true, nil
+					}
+				}
+				revertCompleted = true
```

Comment:
I think this check doesn't work, if a boot order somewhere in the middle is removed. If we stored in `DefaultBootOrder` `[a,b,c]` and b is removed `b != c` will always be true for `idx = 1`.